### ISSUE TITLE
Incomplete chunks received when heading for 1.0/containers?recursion=1

### ISSUE THREAD

**Issue by geodb27** [lxc/lxd#5874]
People hi !

# Required information

 * Distribution: ubuntu
 * Distribution version:  18.04
   * Kernel version: 5.0.0-16-generic
   * LXC version: 3.14
   * LXD version: 3.14
   * Storage backend in use: ceph

# Issue description

When getting, directly from a browser (I've tested with all latest firefox, chrome and chromium on kubuntu), 1.0/containers?recursion=1, I suspect the browser to correct bogus chunks send by the lxd daemon. However, if you go through an apache (2.4) server configured as a proxy between the host and the lxd daemon, apache 2 dislikes the chunks send by the lxd daemon at some point and raises an exception : 
[proxy_http:error] [pid 28309:tid 139831099250432] (70008)Partial results are valid but processing is incomplete: (...) [AH01110] (...).

At first I thought that it was caused by proxy request size limited to 8192 o by default, but I've raised it to 32k and still get the same error.

**Comment by stgraber** [comment]
Does curl report any problem when performing the same request directly against LXD?

Is the reverse proxy with apache running on the same machine as LXD or is the network between the two possibly the cause of this (most results I could find on Google end up with bad network).

**Comment by stgraber** [comment]
Running through some chunk debugging here, I do see a large chunk (11056) followed by the expected zero-length end chunk.

**Comment by geodb27** [comment]
Thanks for your kind ansqer Stéphane, as always. For now, all information I can give you are the following : 
- The apache2 proxy is setup on my laptop. I've made the apache2 user member of group lxd so as to get a certificate I fed to all lxd instance to trust. This is to allow me to go on with something like : http://localhost/lxd_host/1.0/containers?recursion=1 from my own laptop. lxd_host/1.0/containers is "translated" my apache2 proxy module to https://lxc_host:8443/1.0/containers (and so on, the base is on 1.0/.
- This said, I have my own lxd instance on my laptop (so that http://localhost/1.0 is proxied to https://localhost:8443/1.0 so I have both local and remote lxd instances behind my apache2 proxy.
- Last, my network is from one end to the other a 10Gb one, and I have never seen any problem with it. 
However, I'm not abble now to test from curl (I'm at home and not at work). I'll test with curl tomorow.

Here is what led me to think that chunks are incomplete : I have only problems with lxd machines that have so much containers that containers?recursion=1 sends more than 8192 bytes. When less, I have no problem at all. But the failure occurs mainly just after this apache2 limit (that I increased to 32k). If I remember correctly, the failure appeared around 8300. That's why I thought about a malformed chunck at some time. 
The fact is that these errors occur when launched by a XMLHttpRequest. I did many loads on a separate tab, and it just worked fine... That's also what led me to think about what I would freely call "corrupted chunks", because browser are well known to somehow correct malformed responses whenever they can. I don't think this is true to the same extend with the embeded javascript engine.

Hope all these informations will help. Could be caused by the javascript threads also that somehow fall into something quite "unstable" at some point. All this will need more investigations. 

I'll test with curl as you suggest tomorow. Maybe I'll find another way to get the data I need after all. 

Just for your information, in pseudo code, the javascript is something like this : 
Foreach lxd_host do
  get_with_xhtmlrequest (lxd_host/1.0/containers?recursion=true);
done

And the get_with_xhtmlrequest (which are all asyncrhonous) have something to do with onreadystatechange and readystate to fill in a global javascript object handling containers data.

**Comment by stgraber** [comment]
Yeah, in my case I did take a very busy server specifically to get way higher than the 8192, all reported sizes seem correct and the end chunk was present, so hopefully the problem is on apache's side and not in the go http stack :)

**Comment by geodb27** [comment]
Hi again !
Here, I've done the curl try, and the strange thing is that sometime, it works (but I get the same with firefox), sometime I get an error. The error comes more often than when it works however. So, since I get the same error be that with firefox, chrome, chromium or curl, I guess that we can exclude the client problem for once.
- From curl, I get : 

> transfer closed with outstanding read data remaining
> * stopped the pause stream!
> * Closing connection 0
> curl: (18) transfer closed with outstanding read data remaining

- With firefox, the error displayed is quite the same, but don't occur at the same point (here are 3 tries made one after the other on the same url) : 
SyntaxError: JSON.parse: unterminated string at line 1 column 14585 of the JSON data
SyntaxError: JSON.parse: unterminated string at line 1 column 49955 of the JSON data
SyntaxError: JSON.parse: unterminated string at line 1 column 1058 of the JSON data

- With chromium : net::ERR_INCOMPLETE_CHUNKED_ENCODING (size 203B).

I'll continue to investigate, trying different things and using tcpdump if I can, since the transfer between the lxd instance and my machine is using https.

**Comment by geodb27** [comment]
Hi Stephane !
Again thanks for your kind help. Following your thoughts about my problem, and investigating a little more, I finally got it !
The problem was indeed caused by the apache2 mod_proxy and it's default configuration.
What was happening was, from what I understood the following : 
The GET was correctly send to the backend server (the lxd instance).
This last one answered with content-encoding set as chuncked, which was expected. 
The apache reverse-proxy got the chunks, but sent them back to my browser as they were. And, at some time, the chunks were so badly (and unfortunately) segmented that an "End of transmission" was send back to by browser. This one interpreted it (correctly from it's point of view) as "Job done".
In order to avoid this problem, I had to force the transfer-encoding from the reverse proxy to the client to chuncked with this global directive : setenv proxy-sendcl 1
I've tested this for a good amount of time and requests and never had any more error. 
Again, thanks for your kind help !
